### ISSUE TITLE
LSTM  integration results 

### ISSUE THREAD

**Issue by Hdhdgsjawgwi2** [merlresearch/OFENet#1]
import torch
import torch.nn as nn
import torch.optim as optim

class LSTMModule(nn.Module):
    def __init__(self, input_dim, hidden_dim, action_dim, fc_dim, output_dim, num_layers=2, bidirectional=True, dropout=0.2):
        super(LSTMModule, self).__init__()
        
        # LSTM for processing temporal observations
        self.lstm = nn.LSTM(
            input_size=input_dim,    # Observation input dimension
            hidden_size=hidden_dim,  # Hidden state dimension
            num_layers=num_layers,   # Number of stacked LSTM layers
            bidirectional=bidirectional,  # Bidirectional LSTM
            dropout=dropout if num_layers > 1 else 0,  # Dropout applied only if num_layers > 1
            batch_first=True         # Input format: (batch_size, seq_len, input_dim)
        )
        
        # Calculate the output size of the LSTM layer (bidirectional doubles hidden_dim)
        lstm_output_dim = hidden_dim * 2 if bidirectional else hidden_dim
        
        # Fully connected layers for φ₀
        self.fc_φ0 = nn.Sequential(
            nn.Linear(lstm_output_dim, fc_dim),  # Map LSTM output to fc_dim
            nn.ReLU(),
            nn.Linear(fc_dim, fc_dim),  # Second FC layer
            nn.ReLU()
        )
        
        # Fully connected layers for φ₀,a
        self.fc_φ0_a = nn.Sequential(
            nn.Linear(fc_dim + action_dim, fc_dim),  # Combine z_ot with a_t
            nn.ReLU(),
            nn.Linear(fc_dim, output_dim)  # Map to final output
        )
    
    def forward(self, o_t, a_t):
        """
        o_t: Observations (batch_size, seq_len, input_dim)
        a_t: Actions (batch_size, action_dim)
        """
        # φ₀: Process sequential input using LSTM
        lstm_out, _ = self.lstm(o_t)  # LSTM output for all time steps
        h_n = lstm_out[:, -1, :]  # Extract the last time step (batch_size, lstm_output_dim)
        z_ot = self.fc_φ0(h_n)   # Process LSTM output with fully connected layers

        # φ₀,a: Concatenate z_ot and action a_t
        z_concat = torch.cat([z_ot, a_t], dim=-1)  # Concatenate along the feature dimension
        output = self.fc_φ0_a(z_concat)  # Final output
        return output

# Model Configuration
input_dim = 10       # Observation input dimension
hidden_dim = 64      # Hidden size of the LSTM
action_dim = 4       # Action input dimension
fc_dim = 128         # Fully connected layer dimension
output_dim = 1       # Output dimension (e.g., predicted scalar)
num_layers = 2       # Number of LSTM layers
bidirectional = True # Use bidirectional LSTM
dropout = 0.2        # Dropout rate

# Initialize the model
model = LSTMModule(input_dim, hidden_dim, action_dim, fc_dim, output_dim, num_layers, bidirectional, dropout)

# Example input
batch_size = 8
seq_len = 5  # Number of time steps in the sequence
o_t = torch.randn(batch_size, seq_len, input_dim)  # Random observation data
a_t = torch.randn(batch_size, action_dim)          # Random action data
target = torch.randn(batch_size, output_dim)       # Target output

# Loss and optimizer
criterion = nn.MSELoss()  # Example loss function
optimizer = optim.Adam(model.parameters(), lr=0.001)

# Forward pass
output = model(o_t, a_t)
loss = criterion(output, target)

# Backward pass
loss.backward()
optimizer.step()

print(f"Output: {output}")
print(f"Loss: {loss.item()}")
